[13604] Run GHA workflow tests on each PR

radoslawkrzemien opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Added a new workflow that runs previously added GHA workflow tests on each PR (excluding PRs to `staging` and `production`)

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/13604#issuecomment-1724808403


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

radoslawkrzemien committed:
Add workflow

Added the first version of GHA workflow running workflow tests on opening and syncing of PRs

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix

runs -> run

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add trigger

Added workflow_dispatch trigger

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update trigger

Added PR event type of edited

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge pull request #1 from radoslawkrzemien/@radoslawkrzemien/13604-run-workflow-tests-on-pr

Run GHA workflow tests on each PR

radoslawkrzemien committed:
Add step

Added step for Act installation

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add step

Added step for Homebrew installation

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Updated step for Homebrew installation with post install commands

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Added some debug echoes

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Found existing Homebrew setup step

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add step

Added step for setting the ACT_BINARY env variable

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Updated step for setting the ACT_BINARY to use the way proposed by GHA

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Updated step for running tests with debug echoes to get into what is the working dir

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update step

Updated step for running tests by removing debug echoes

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for preDeploy after two of the jobs have been removed

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for testBuild after one of the messages have been updated

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for cherryPick after the way github token is passed to a step has changed

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for platformDeploy after 2 new steps have been added

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for finishReleaseCycle after 2 new steps have been added and the way the token is passed has changed

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for createNewVersion after new secrets have been added and the way the token is passed has changed

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for deploy after new secrets have been added and the way the token is passed has changed

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed tests for lint after new steps have been added

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Run prettier

Run prettier

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add quotes

Add quotes to satisfy linter

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Remove strategy

Removed strategy field from gha workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge branch 'Expensify:main' into main

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into @radoslawkrzemien/13604-run-workflow-tests-on-pr

sobitneupane commented:
I'm not confident in my expertise to thoroughly review this PR. @AndrewGable @dangrous, could you please assign another reviewer (C+) for this PR? Thank you!

dangrous commented:
Sure thing! Or at least I will try. I just did and it assigned me lol

dangrous commented:
welp. I'll ask for a volunteer in Slack

rushatgabhane commented:
hiya! you can assign it to me

rushatgabhane commented:
This is a test that `workflow-test` would run locally. Right??

radoslawkrzemien commented:
The new workflow executes on PR events, when contents of `.github` directory changes, runs setup and executes `npm run workflow-test`. This in turn checks if the environment is set up correctly and runs all the tests located in `workflow_tests` directory using `jest`

rushatgabhane commented on .github/workflows/testGithubActionsWorkflows.yml:
```diff
@@ -0,0 +1,35 @@
+  pull_request:
+    types: [opened, reopened, edited, synchronize]
+    branches-ignore: [staging, production]
+    paths: ['.github']
```
This workflow should have run in this PR also. But I don't see it here.

I think it might be our path?
```suggestion
    paths: ['.github/**']
```


<img width="868" alt="image" src="https://github.com/Expensify/App/assets/29673073/6f5a9cd2-e15e-42e2-bf31-36c416546743">

radoslawkrzemien replied:
Yes, you're probably right

radoslawkrzemien replied:
Updated

radoslawkrzemien committed:
Add wildcards

Updated the paths field in the workflow with the wildcard to match all files in .github directory and subdirectories

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Move actor definition

Declares and defines actor value within the test

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
Not sure why these particular tests fail in GHA as all tests pass locally, need to investigate it some more

radoslawkrzemien committed:
Merge branch 'Expensify:main' into main

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into @radoslawkrzemien/13604-run-workflow-tests-on-pr

# Conflicts:
#	.github/workflows/deploy.yml
#	.github/workflows/finishReleaseCycle.yml

radoslawkrzemien committed:
Fix workflow tests

Fixed platformDeploy workflow tests after recent changes in upstream

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix workflow tests

Fixed authorChecklist workflow tests after recent changes in upstream

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix workflow tests

Fixed deployBlocker workflow tests after recent changes in upstream

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Run prettier

Run prettier

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
Looks good now

rushatgabhane reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

rushatgabhane reviewed (approved):
LGTM

dangrous reviewed (dismissed):
One very not a blocker comment, otherwise looks good. Double checked this still runs well locally as well as on GH. Will let @AndrewGable take a last look!

radoslawkrzemien committed:
Type git with lowercase

Changed step names containing git to have it in lowercase

See: https://github.com/Expensify/App/issues/13604

dangrous reviewed (approved):
All yours, @AndrewGable !

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7031929898) to production by https://github.com/mountiny in version: 1.4.4-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
